mm/kfence/core.c

```diff
@@ -86,20 +86,49 @@ static inline bool kfence_unprotect(unsigned long addr)
 	return !KFENCE_WARN_ON(!kfence_change_page_prot(ALIGN_DOWN(addr, PAGE_SIZE), false));
 }
 
-static inline struct kfence_metadata *kfence_addr_to_metadata(unsigned long addr)
+static inline struct kfence_metadata *addr_to_metadata(unsigned long addr)
 {
 	long index;
 
+	/* The checks do not affect performance; only called from slow-paths. */
+
 	if (!is_kfence_addr((void *)addr))
 		return NULL;
 
-	index = ((addr - (unsigned long)__kfence_pool) / PAGE_SIZE / 2) - 1;
+	/*
+	 * May be an invalid index if called with an address at the edge of
+	 * __kfence_pool, in which case we would report an "invalid access"
+	 * error.
+	 */
+	index = ((addr - (unsigned long)__kfence_pool) / (PAGE_SIZE * 2)) - 1;
 	if (index < 0 || index >= CONFIG_KFENCE_NUM_OBJECTS)
```

Comment:
This branch is unneeded now, right?

Reply:
Not quite, we can still get an address that is the first (reserved) page that'll pass is_kfence_addr. In this case, the index will be -1.

I added a test for invalid accesses which exercises this.